File: exercism/kotlin/difference-of-squares/src/main/kotlin/Squares.kt
class Squares(val n: Int) {
  
  fun squareOfSum(): Int = (n + 1) * (n + 1) * n * n / 4

  fun sumOfSquares(): Int = (2 * n + 1) * (n + 1) * n / 6

  fun difference(): Int = squareOfSum() - sumOfSquares()
}

File: exercism/haskell/difference-of-squares/src/Squares.hs
module Squares (sumOfSquares, squareOfSums, difference) where

powerSum :: Integral a => a -> a -> a -> a
powerSum a b n = (^a) $ sum $ map (^b) [1..n]

sumOfSquares :: Integral a => a -> a
sumOfSquares = powerSum 1 2

squareOfSums :: Integral a => a -> a
squareOfSums = powerSum 2 1

difference :: Integral a => a -> a
difference n = squareOfSums n - sumOfSquares n
